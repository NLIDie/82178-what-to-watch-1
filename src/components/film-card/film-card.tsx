import * as React from "react";
import {Link} from "react-router-dom";

// Components
import {VideoPlayer} from "_Components/video-player/video-player";

// Types
import {TProps} from "./film-card.types";

export function FilmCard(props: TProps) {
  const {id, title, poster, previewVideo} = props;

  const refTimeout = React.useRef(null);

  const handleMouseOver = (event: React.SyntheticEvent<HTMLVideoElement>) => {
    const videoPlayer = event.currentTarget;

    if (videoPlayer) {
      refTimeout.current = setTimeout(() => {
        const promise = videoPlayer.play();

        if (promise !== undefined) {
          promise.catch(() => {}).then(() => {});
        }
      }, 1000);
    }
  };

  const handleMouseOut = (event: React.SyntheticEvent<HTMLVideoElement>) => {
    const videoPlayer = event.currentTarget;

    if (videoPlayer) {
      videoPlayer.load();
    }

    clearTimeout(refTimeout.current);
  };

  return (
    <article className="small-movie-card catalog__movies-card">
      <Link className="small-movie-card__link" to={`/film/${id}`}>
        <VideoPlayer
          src={previewVideo}
          poster={poster}
          onMouseOut={handleMouseOut}
          onMouseOver={handleMouseOver}
        />
      </Link>

      <h3 className="small-movie-card__title">
        <Link className="small-movie-card__link" to={`/film/${id}`}>
          {title}
        </Link>
      </h3>
    </article>
  );
}
