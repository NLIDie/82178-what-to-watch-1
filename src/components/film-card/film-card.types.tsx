import * as React from "react";

export type TProps = {
  readonly id: number;
  readonly title: string;
  readonly preview: string;
  readonly onPlay: (
    event: React.SyntheticEvent<HTMLElement>,
    filmId: number
  ) => void;
};
