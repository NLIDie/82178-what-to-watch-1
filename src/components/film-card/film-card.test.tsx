import * as React from "react";
import {create} from "react-test-renderer";

// Components
import {FilmCard} from "./film-card";

jest.mock(`react-router-dom`, () => ({
  BrowserRouter: ({children}) => <div>{children}</div>,
  Route: ({children}) => <div>{children}</div>,
  Link: ({children}) => <a>{children}</a>
}));

describe(`Components/FilmCard`, () => {
  it(`Correct renderer the component`, () => {
    const tree = create(
        <FilmCard
          id={1}
          title="Movie title"
          preview="empty"
          onPlay={jest.fn()}
        />
    );

    expect(tree).toMatchSnapshot();
  });
});
