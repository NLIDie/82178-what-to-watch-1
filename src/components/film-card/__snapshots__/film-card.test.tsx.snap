// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/FilmCard Correct renderer the component 1`] = `
<article
  className="small-movie-card catalog__movies-card"
>
  <button
    className="small-movie-card__play-btn"
    onClick={[Function]}
    type="button"
  >
    Play
  </button>
  <div
    className="small-movie-card__image"
  >
    <img
      alt="Movie title"
      height="175"
      src="empty"
      width="280"
    />
  </div>
  <h3
    className="small-movie-card__title"
  >
    <a>
      Movie title
    </a>
  </h3>
</article>
`;
